docs/content/sips/xxx-spin-plugins.md

```diff
@@ -0,0 +1,154 @@
+title = "SIP xxx - Plugin System for Spin"
+template = "main"
+date = "2022-08-09T13:22:30Z"
+---
+
+Summary: Plugin system for spin.
+
+Owner: karthik.ganeshram@fermyon.com
+
+Created: August 9, 2022 
+
+## Background 
+
+As the functionality fo spin gets extended, there will be a point where every feature will not be required/used by every user. An example of this would be different triggers for spin along with possible subcommands to add more functionality to spin. Therefore it would make sense to be able to add features as plugins allowing for addition of only the required features witout modifying the spin binary.
+
+## Proposal
+
+Create a new subcommand for spin called `spin plugin` which will have two further subcommands namely
+
+- `install`
+- `uninstall`
+
+### Types of plugins
+
+![Classification of types of plugins](https://i.imgur.com/Fo7EGPQ.png)
+
+- Standalone Binary Plugins (external subcommands) - These work to add features to spin and do not need any use of the spin binary.
+- Internal Spin Plugins - These are plugins like triggers that are used by the spin application. 
+- Spin dependant User Plugins - These user invoked plugins require some functionality provided by spin.
+
+### Proposed workflow
+
+A plugin can be installed using the following command.
+
+```bash
+$ spin plugin install <name_of_plugin>
+The following $plugin from <source_of_plugin> will be installed. 
+Make sure you trust this source! Continue? <y/N> 
+```
+
+This installs the plugin to a to a spin managed directory based on the operating system.
+
+In the case of standaone binary plugins and spin dependant user plugins, the plugin can be invoked as 
+
+```bash
+$ spin $plugin <args_to_plugin>
+```
+
+In the case of spin Dependant Internal Plugins, the user does not directly invoke the command, these plugins are directly invoked by spin where required (i.e) like triggers based on the application manifest. [(Spin Trigger Executor)](https://spin.fermyon.dev/sips/003-trigger-executors.md)
+
+To uninstall plugins, the following command can be used.
+
+```
+spin uninstall $plugin
+```
+
+This will uninstall the executable from the spin managed directory.
+
+
+### Execution of plugins
+
+The execution of the plugins works differently based on the type of plugin.
+
+- **Standalone Binary Plugin** 
+  - Once the plugin is invoked using `spin $plugin <args_to_plugin>`, the external subcommand is invoked as `spin-$plugin <args>`.
+- **Spin Dependant User Plugins** 
+  - Spin first provides all the functionality available in it required by the plugin and then proceeds to call the plugin along with the required parameters.
+
+- **Spin Dependant Internal Plugins**
+  - These plugins cannot be directly invoked by the user as it is used by spin internally. Therefore, only spin invokes these.  
+
+## Proposed implementation of `spin plugin`
+
+### The `install` subcommand
```

Comment:
Yes, this sounds like a good idea. I will work on it.

Reply:
It would be very nice (perhaps in a second iteration) to have something equivalent to `spin plugin update --all`. I think the expected behavior differs from templates in that you pretty much always want the latest version of a plugin.

Reply:
Makes sense. Should an update do a check to see if an update is needed or always overwrite, even if it is bringing down the same version? I am not sure what the best way to do a version check would be, unless we save the current plugin manifest in the directory of the plugin and query the version there and upstream to see if there is a newer version.

Reply:
Yeah I think this would require storing information about anything installed via `spin plugin install` somewhere locally. I would probably lean towards a e.g. `plugins.json` in the same parent directory of `plugins/`